lib/jsonrpccxx/client.hpp

```diff
@@ -1,92 +1,126 @@
+  id_type id;
+  json result;
+};
 
-    template <typename T>
-    T CallMethod(const id_type &id, const std::string &name) { return call_method(id, name, json::object()).result.get<T>(); }
-     template <typename T>
-    T CallMethod(const id_type &id, const std::string &name, const positional_parameter &params) { return call_method(id, name, params).result.get<T>(); }
-    template <typename T>
-    T CallMethodNamed(const id_type &id, const std::string &name, const named_parameter &params = {}) { return call_method(id, name, params).result.get<T>(); }
+class JsonRpcClient {
+ public:
+  JsonRpcClient(IClientConnector &connector, version v)
+      : connector(connector), v(v) {}
+  virtual ~JsonRpcClient() = default;
 
-    void CallNotification(const std::string &name, const positional_parameter &params = {}) { call_notification(name, params); }
-    void CallNotificationNamed(const std::string &name, const named_parameter &params = {}) { call_notification(name, params); }
+  template <typename T>
+  T CallMethod(const id_type &id, const std::string &name) {
+    return call_method(id, name, json::object()).result.get<T>();
+  }
 
-  protected:
-    IClientConnector &connector;
+  template <typename T>
+  T CallMethod(const id_type &id, const std::string &name,
+               const positional_parameter &params) {
+    return call_method(id, name, params).result.get<T>();
+  }
 
-  private:
-    version v;
+  template <typename T>
+  T CallMethodNamed(const id_type &id, const std::string &name,
+                    const named_parameter &params = {}) {
+    return call_method(id, name, params).result.get<T>();
+  }
 
-    JsonRpcResponse call_method(const id_type &id, const std::string &name, const json &params) {
-      json j = {{"method", name}};
-      if (std::get_if<int>(&id) != nullptr) {
-        j["id"] = std::get<int>(id);
-      } else {
-        j["id"] = std::get<std::string>(id);
-      }
-      if (v == version::v2) {
-        j["jsonrpc"] = "2.0";
-      }
-      if (!params.empty() && !params.is_null()) {
-        j["params"] = params;
-      } else if (params.is_array()) {
-        j["params"] = params;
-      } else if (v == version::v1) {
-        j["params"] = nullptr;
+  nlohmann::json CallMethodNamedRetJson(const id_type &id,
+                                        const std::string &name,
+                                        const named_parameter &params = {}) {
+    return call_method(id, name, params).result;
+  }
+
+  void CallNotification(const std::string &name,
+                        const positional_parameter &params = {}) {
+    call_notification(name, params);
+  }
+
+  void CallNotificationNamed(const std::string &name,
+                             const named_parameter &params = {}) {
+    call_notification(name, params);
+  }
+
+ protected:
+  IClientConnector &connector;
+
+ private:
+  version v;
+
+  JsonRpcResponse call_method(const id_type &id, const std::string &name,
```

Comment:
we are doing here direct copy ? Since struct do not have move constructor ?

Reply:
He is talking about returning the `JsonRpcResponse` struct. Structs and classes are nearly identical (apart from the default state of the accesses specifiers) so there are move constructors for struct also (if there are no user-defined constructors for that struct - but here it is not that case).

Reply:
Ok, so If there are any improvements in casper-cpp-sdk you want to make, raise jira tickets and we will handle it in future. For now I just wanna merge my changes to unblock removal of casper-client and not waste too much time on that repo :)